src/com/intellij/stats/ngram/NGramFileBasedIndex.kt

```diff
@@ -0,0 +1,91 @@
+package com.intellij.stats.ngram
+
+import com.intellij.openapi.module.ModuleUtil
+import com.intellij.psi.PsiElement
+import com.intellij.psi.PsiFile
+import com.intellij.psi.PsiRecursiveElementVisitor
+import com.intellij.psi.search.GlobalSearchScope
+import com.intellij.util.indexing.*
+import com.intellij.util.io.DataExternalizer
+import com.intellij.util.io.KeyDescriptor
+import java.io.DataInput
+import java.io.DataOutput
+
+class NGramFileBasedIndex : FileBasedIndexExtension<NGram, Int>() {
+    val inputFilter = DefaultFileTypeSpecificInputFilter(*NGramElementProvider.getSupportedFileTypes().toTypedArray())
+
+    override fun getValueExternalizer(): DataExternalizer<Int> {
+        return IntDataExternalizer
+    }
+
+    override fun getName(): ID<NGram, Int> {
+        return KEY
+    }
+
+    override fun getVersion(): Int {
+        return INDEX_VERSION
+    }
+
+    override fun dependsOnFileContent(): Boolean {
+        return true
+    }
+
+    override fun getIndexer(): DataIndexer<NGram, Int, FileContent> {
+        return NGramIndexer
+    }
+
+    override fun getInputFilter(): FileBasedIndex.InputFilter {
+        return inputFilter
+    }
+
+    override fun getKeyDescriptor(): KeyDescriptor<NGram> {
+        return NGramKeyDescriptor
+    }
+
+    companion object {
+        val KEY = ID.create<NGram, Int>("ngram.index")
+        const val INDEX_VERSION = 1
+
+        fun getNumberOfOccurrences(key: NGram, scope: GlobalSearchScope): Int {
+            return FileBasedIndex.getInstance().getValues(KEY, key, scope).sum()
+        }
+    }
+}
+
+private object NGramIndexer : DataIndexer<NGram, Int, FileContent> {
+    override fun map(fileContent: FileContent): MutableMap<NGram, Int> {
+        if (ModuleUtil.findModuleForFile(fileContent.file, fileContent.project) == null) {
+            return HashMap()
```

Comment:
`emptyMap()` could be used. But at first, replace return type with `Map<NGram, Int>`